### PR TITLE
[Torch Ranker Agent] Fixed candidate fixes

### PR DESCRIPTION
**Pull request by emilydinan**
- be able to optionally encode and cache fixed candidates (models must implement how to encode them, but TRA handles loading and saving)
- implement this for the bert biencoder ranker

### CODE REVIEW

File: parlai/agents/bert_ranker/bi_encoder_ranker.py [hunk 1]
```diff
@@ -93,6 +90,12 @@ def score_candidates(self, batch, cand_vecs):
         embedding_ctxt, _ = self.model(
             token_idx_ctxt, segment_idx_ctxt, mask_ctxt,
             None, None, None)
+
+        # evaluating a fixed set of candidates
+        if (hasattr(self, 'fixed_candidate_encs') and self.fixed_candidate_encs
```
Comment (stephenroller) [inline parlai/agents/bert_ranker/bi_encoder_ranker.py]: imho
```python
if (hasattr(self, 'fixed_candidate_encs') and
        self.fixed_candidate_encs is not None):
```

File: parlai/core/torch_ranker_agent.py [hunk 2]
```diff
@@ -5,6 +5,7 @@
 # LICENSE file in the root directory of this source tree.
 
 import os
+from tqdm import tqdm
```
Comment (stephenroller) [inline parlai/core/torch_ranker_agent.py]: <3

File: parlai/core/torch_ranker_agent.py [hunk 3]
```diff
@@ -26,40 +27,42 @@ def add_cmdline_args(cls, argparser):
...
             help='Get predictions and calculate mean rank during the train '
-                 'step. Turning this on may slow down training.'
-        )
+                 'step. Turning this on may slow down training.')
         agent.add_argument(
             '--ignore-bad-candidates', type='bool', default=False,
             help='Ignore examples for which the label is not present in the '
-                 'label candidates. Default behavior results in RuntimeError. '
-        )
+                 'label candidates. Default behavior results in RuntimeError. ')
```
Comment (stephenroller) [inline parlai/core/torch_ranker_agent.py]: lol are you and i having a parenthesis war?

File: parlai/core/torch_ranker_agent.py [hunk 4]
```diff
@@ -547,5 +588,5 @@ def vectorize_fixed_candidates(self, cands_batch):
         A child class may choose to overwrite this method to perform vectorization as
         well as encoding if so desired.
         """
-        return [self._vectorize_text(cand, truncate=self.truncate, truncate_left=False)
-                for cand in cands_batch]
+        return [self._vectorize_text(cand, truncate=self.label_truncate,
```
Comment (stephenroller) [inline parlai/core/torch_ranker_agent.py]: :-O it was more readable before

File: parlai/core/torch_ranker_agent.py (hunk 3 above)
Comment (emilydinan) [reply parlai/core/torch_ranker_agent.py]: haha i was just making it match the other args but i can make it do whatever

Comment (stephenroller) [reply parlai/core/torch_ranker_agent.py]: consistency with the file is better. your change is good

**Comment by jaseweston** [comment]
gonna merge this and assumes work as i need to do some exp with it
if there's more wrong let's do in anther PR

**Pull request merged**